File: backend/services/bdnb.service.ts
/**
 * BDNB Building Data Service — Server-side fetcher
 */
import type { BuildingData } from '../models/schema.js';

function constructionPeriod(year: number | null): BuildingData['constructionPeriod'] {
  if (!year) return null;
  if (year < 1915) return '<1915';
  if (year <= 1948) return '1915_1948';
  if (year <= 1974) return '1949_1974';
  if (year <= 2000) return '1975_2000';
  if (year <= 2012) return '2001_2012';
  if (year <= 2021) return '2013_2021';
  return '>2021';
}

export async function fetchBuildingByBanId(banId: string): Promise<BuildingData> {
  const deptCode = banId.slice(0, 2);
  const empty: BuildingData = {
    builtYear: null, constructionPeriod: null, surfaceUtile: null, surfaceEmprise: null,
    levels: null, height: null, dpeClass: null, energyConsumption: null,
    emissionGes: null, wallMaterial: null, roofMaterial: null, heatingType: null,
    usageType: null, nbLogements: null, departmentCode: deptCode,
    nbLogementsRnc: null, clayExposure: null, altitudeSolMean: null,
    heatingEnergyType: null, parcelIds: null, quartierPrioritaire: null,
    zonePatrimoniale: null, footprint: null,
  };

  try {
    const url = `https://api.bdnb.io/v1/bdnb/donnees/rel_batiment_groupe_adresse?cle_interop_adr=eq.${banId}&select=batiment_groupe_id`;
    const res = await fetch(url, { headers: { Accept: 'application/json' } });
    if (res.ok) {
      const relData = await res.json();
      const groupIds: string[] = (Array.isArray(relData) ? relData : [])
        .map((r: any) => r.batiment_groupe_id)
        .filter(Boolean);

      if (groupIds.length > 0) {
        const idsParam = groupIds.map((id: string) => `"${id}"`).join(',');
        const bdgUrl = `https://api.bdnb.io/v1/bdnb/donnees/batiment_groupe_complet?batiment_groupe_id=in.(${idsParam})`;
        const bdgRes = await fetch(bdgUrl, { headers: { Accept: 'application/json' } });
        if (bdgRes.ok) {
          const bdgData = await bdgRes.json();
          const arr = Array.isArray(bdgData) ? bdgData : bdgData?.features || [];
          if (arr.length > 0) {
            const props = arr[0].properties || arr[0];
            const year = props.annee_construction ?? null;
            return {
              builtYear: year,
              constructionPeriod: constructionPeriod(year),
              surfaceUtile: props.surface_habitable ?? null,
              surfaceEmprise: props.surface_emprise_sol ?? null,
              levels: props.nb_niveau ?? null,
              height: props.hauteur_mean ?? props.hauteur ?? null,
              dpeClass: props.classe_bilan_dpe ?? null,
              energyConsumption: props.conso_energie ?? null,
              emissionGes: props.emission_ges ?? null,
              wallMaterial: props.mat_mur_txt && props.mat_mur_txt !== 'INDETERMINE' ? props.mat_mur_txt : null,
              roofMaterial: props.mat_toit_txt && props.mat_toit_txt !== 'INDETERMINE' ? props.mat_toit_txt : null,
              heatingType: props.etat_chauffage_txt && props.etat_chauffage_txt !== 'INDETERMINE' ? props.etat_chauffage_txt : null,
              usageType: props.usage_principal_bdnb_open ?? null,
              nbLogements: props.nb_logements ?? null,
              departmentCode: deptCode,
              nbLogementsRnc: props.nb_log ?? null,
              clayExposure: props.alea_argile && props.alea_argile !== 'INDETERMINE' ? props.alea_argile : null,
              altitudeSolMean: props.altitude_sol_mean ?? null,
              heatingEnergyType: props.type_energie_chauffage ?? null,
              parcelIds: Array.isArray(props.l_parcelle_id) ? props.l_parcelle_id : null,
              quartierPrioritaire: props.quartier_prioritaire === true || props.quartier_prioritaire === 'true',
              zonePatrimoniale: props.zone_plu_bati_patrimonial ?? null,
              footprint: arr[0]?.geometry || null,
            };
          }
        }
      }
    }
  } catch {
    // Ignore error and return empty fallback
  }

  return empty;
}

File: backend/services/bdnb_service.py
"""
BDNB Building Data Service — Server-side fetcher
=================================================
Direct port of `backend/services/bdnb.service.ts`.

Fetches building characteristics from the Base Nationale des Bâtiments (BDNB).
"""

from __future__ import annotations

from typing import Any

import httpx

CLIENT_TIMEOUT = 10.0


def _construction_period(year: int | None) -> str | None:
    if not year:
        return None
    if year < 1915:
        return "<1915"
    if year <= 1948:
        return "1915_1948"
    if year <= 1974:
        return "1949_1974"
    if year <= 2000:
        return "1975_2000"
    if year <= 2012:
        return "2001_2012"
    if year <= 2021:
        return "2013_2021"
    return ">2021"


def _empty_building(dept_code: str) -> dict[str, Any]:
    return {
        "built_year": None,
        "construction_period": None,
        "surface_utile": None,
        "surface_emprise": None,
        "levels": None,
        "height": None,
        "dpe_class": None,
        "energy_consumption": None,
        "emission_ges": None,
        "wall_material": None,
        "roof_material": None,
        "heating_type": None,
        "usage_type": None,
        "nb_logements": None,
        "department_code": dept_code,
        "nb_logements_rnc": None,
        "clay_exposure": None,
        "altitude_sol_mean": None,
        "heating_energy_type": None,
        "parcel_ids": None,
        "quartier_prioritaire": None,
        "zone_patrimoniale": None,
        "footprint": None,
    }


class BdnbService:
    """Fetch building data from BDNB API."""

    @staticmethod
    async def fetch_building_by_ban_id(ban_id: str) -> dict[str, Any]:
        dept_code = ban_id[:2]
        empty = _empty_building(dept_code)

        try:
            async with httpx.AsyncClient(timeout=CLIENT_TIMEOUT) as client:
                # Step 1: resolve ban_id → batiment_groupe_id
                url = (
                    f"https://api.bdnb.io/v1/bdnb/donnees/rel_batiment_groupe_adresse"
                    f"?cle_interop_adr=eq.{ban_id}&select=batiment_groupe_id"
                )
                res = await client.get(url, headers={"Accept": "application/json"})
                if not res.is_success:
                    return empty

                rel_data = res.json()
                rel_list = rel_data if isinstance(rel_data, list) else []
                group_ids = [
                    r["batiment_groupe_id"]
                    for r in rel_list
                    if r.get("batiment_groupe_id")
                ]
                if not group_ids:
                    return empty

                # Step 2: fetch building details
                ids_param = ",".join(f'"{gid}"' for gid in group_ids)
                bdg_url = (
                    f"https://api.bdnb.io/v1/bdnb/donnees/batiment_groupe_complet"
                    f"?batiment_groupe_id=in.({ids_param})"
                )
                bdg_res = await client.get(bdg_url, headers={"Accept": "application/json"})
                if not bdg_res.is_success:
                    return empty

                bdg_data = bdg_res.json()
                arr = bdg_data if isinstance(bdg_data, list) else bdg_data.get("features", [])
                if not arr:
                    return empty

                props = arr[0].get("properties", arr[0])
                year = props.get("annee_construction") or None

                return {
                    "built_year": year,
                    "construction_period": _construction_period(year),
                    "surface_utile": props.get("surface_habitable"),
                    "surface_emprise": props.get("surface_emprise_sol"),
                    "levels": props.get("nb_niveau"),
                    "height": props.get("hauteur_mean") or props.get("hauteur"),
                    "dpe_class": props.get("classe_bilan_dpe"),
                    "energy_consumption": props.get("conso_energie"),
                    "emission_ges": props.get("emission_ges"),
                    "wall_material": (
                        props["mat_mur_txt"]
                        if props.get("mat_mur_txt") and props["mat_mur_txt"] != "INDETERMINE"
                        else None
                    ),
                    "roof_material": (
                        props["mat_toit_txt"]
                        if props.get("mat_toit_txt") and props["mat_toit_txt"] != "INDETERMINE"
                        else None
                    ),
                    "heating_type": (
                        props["etat_chauffage_txt"]
                        if props.get("etat_chauffage_txt") and props["etat_chauffage_txt"] != "INDETERMINE"
                        else None
                    ),
                    "usage_type": props.get("usage_principal_bdnb_open"),
                    "nb_logements": props.get("nb_logements"),
                    "department_code": dept_code,
                    "nb_logements_rnc": props.get("nb_log"),
                    "clay_exposure": (
                        props["alea_argile"]
                        if props.get("alea_argile") and props["alea_argile"] != "INDETERMINE"
                        else None
                    ),
                    "altitude_sol_mean": props.get("altitude_sol_mean"),
                    "heating_energy_type": props.get("type_energie_chauffage"),
                    "parcel_ids": (
                        props["l_parcelle_id"]
                        if isinstance(props.get("l_parcelle_id"), list)
                        else None
                    ),
                    "quartier_prioritaire": (
                        props.get("quartier_prioritaire") in (True, "true")
                    ),
                    "zone_patrimoniale": props.get("zone_plu_bati_patrimonial"),
                    "footprint": arr[0].get("geometry"),
                }
        except Exception:
            return empty
